bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/WatchdogInterceptor.java

```diff
@@ -0,0 +1,208 @@
+
+  private ScheduledFuture<?> future;
+
+
+  public WatchdogInterceptor(Set<MethodDescriptor<?, ?>> watchedMethods,
+      ScheduledExecutorService executor,
+      ApiClock clock, Duration waitTimeout, Duration idleTimeout) {
+    this.watchedMethods = watchedMethods;
+    this.executor = executor;
+    this.clock = Preconditions.checkNotNull(clock, "clock can't be null");
+    this.waitTimeout = Preconditions.checkNotNull(waitTimeout, "waitTimeout can't be null");
+    this.idleTimeout = Preconditions.checkNotNull(idleTimeout, "idleTimeout can't be null");
+  }
+
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+      MethodDescriptor<ReqT, RespT> methodDescriptor, CallOptions callOptions, Channel channel) {
+
+    ClientCall<ReqT, RespT> call = channel.newCall(methodDescriptor, callOptions);
+    if (watchedMethods.contains(methodDescriptor)) {
+      call = new WatchedCall<>(call);
+    }
+
+    return call;
+  }
+
+  public void start() {
+    future = executor.scheduleAtFixedRate(new Runnable() {
+      @Override
+      public void run() {
+        abortStaleStreams();
+      }
+    }, 0, waitTimeout.dividedBy(2).getMillis(), TimeUnit.MILLISECONDS);
+  }
+
+  public void stop() {
+    future.cancel(true);
+  }
+
+
+  public void abortStaleStreams() {
+    Iterator<Entry<WatchedCall<?,?>, Object>> it = openStreams.entrySet().iterator();
+
+    while (it.hasNext()) {
+      WatchedCall<?,?> stream = it.next().getKey();
+      if (stream.cancelIfStale()) {
+        it.remove();
+      }
+    }
+  }
+
+  enum State {
+    /** Stream has been started, but doesn't have any outstanding requests. */
+    IDLE,
+    /** Stream is awaiting a response from upstream. */
+    WAITING,
+    /**
+     * Stream received a response from upstream, and is awaiting outerResponseObserver processing.
+     */
+    DELIVERING
+  }
+
+  class WatchedCall<ReqT, RespT> extends SimpleForwardingClientCall<ReqT, RespT> {
+    private final Object lock = new Object();
+
+    @GuardedBy("lock")
+    private State state;
+
+    @GuardedBy("lock")
+    private int pendingCount = 0;
+
+    @GuardedBy("lock")
+    private long lastActivityAt = clock.millisTime();
+
+    WatchedCall(ClientCall<ReqT, RespT> delegate) {
+      super(delegate);
+    }
+
+    @Override
+    public void start(Listener<RespT> listener, Metadata metadata) {
+      ClientCall<ReqT, RespT> call = delegate();
+
+      synchronized (lock) {
+        Preconditions.checkState(state == null, "Already started");
+        state = (pendingCount == 0) ? State.IDLE : State.WAITING;
+        lastActivityAt = clock.millisTime();
+      }
+
+      openStreams.put(this, PRESENT);
+
+      call.start(new SimpleForwardingClientCallListener<RespT>(listener) {
+        @Override
+        public void onMessage(RespT message) {
+          synchronized (lock) {
+            state = State.DELIVERING;
+          }
+
+          super.onMessage(message);
+
+          synchronized (lock) {
```

Comment:
please add this in a try-finally block, just in case.  It might also make sense to do this in a catch block.